server/container_create.go

```diff
@@ -240,58 +240,93 @@ func buildOCIProcessArgs(containerKubeConfig *pb.ContainerConfig, imageOCIConfig
-		} else {
-			// Case 2: run as username is set by kubelet
-			userName := sc.GetRunAsUsername()
-			if userName != "" {
-				containerUser = userName
-			} else {
-				// Case 3: get user from image config
-				if imageConfig != nil {
-					imageUser := imageConfig.Config.User
-					if imageUser != "" {
-						containerUser = imageUser
-					}
-				}
-			}
-		}
-		if sc.GetRunAsUser() == nil && sc.GetRunAsUsername() == "" && sc.GetRunAsGroup() != nil {
-			return fmt.Errorf("RunAsGroup should be specified only with RunAsUser or RunAsUsername")
-		}
-		groupstr := strconv.FormatInt(sc.GetRunAsGroup().GetValue(), 10)
-		if groupstr != "" {
-			containerUser = containerUser + ":" + groupstr
-		}
+	if sc == nil {
+		return nil
+	}
+	imageUser := ""
+	if imageConfig != nil {
+		imageUser = imageConfig.Config.User
+	}
+	containerUser, err := generateUserString(
+		sc.GetRunAsUsername(),
+		imageUser,
+		sc.GetRunAsUser(),
+		sc.GetRunAsGroup(),
+	)
+	if err != nil {
+		return err
+	}
 
-		logrus.Debugf("CONTAINER USER: %+v", containerUser)
+	logrus.Debugf("CONTAINER USER: %+v", containerUser)
 
-		// Add uid, gid and groups from user
-		uid, gid, addGroups, err1 := getUserInfo(rootfs, containerUser)
-		if err1 != nil {
-			return err1
-		}
+	// Add uid, gid and groups from user
+	uid, gid, addGroups, err := getUserInfo(rootfs, containerUser)
+	if err != nil {
+		return err
+	}
 
-		logrus.Debugf("UID: %v, GID: %v, Groups: %+v", uid, gid, addGroups)
-		specgen.SetProcessUID(uid)
-		specgen.SetProcessGID(gid)
-		for _, group := range addGroups {
-			specgen.AddProcessAdditionalGid(group)
-		}
+	logrus.Debugf("UID: %v, GID: %v, Groups: %+v", uid, gid, addGroups)
+	specgen.SetProcessUID(uid)
+	specgen.SetProcessGID(gid)
 
-		// Add groups from CRI
-		groups := sc.GetSupplementalGroups()
-		for _, group := range groups {
-			specgen.AddProcessAdditionalGid(uint32(group))
-		}
+	if sc.GetRunAsUsername() != "" {
+		containerUser = sc.GetRunAsUsername()
+	} else {
+		containerUser = strconv.FormatInt(sc.GetRunAsUser().GetValue(), 10)
+	}
+
+	_, _, addGroups2, err := getUserInfo(rootfs, containerUser)
+	if err != nil {
+		return err
+	}
+
+	for _, group := range addGroups {
+		specgen.AddProcessAdditionalGid(group)
+	}
+	for _, group := range addGroups2 {
+		specgen.AddProcessAdditionalGid(group)
+	}
+
+	// Add groups from CRI
+	groups := sc.GetSupplementalGroups()
+	for _, group := range groups {
+		specgen.AddProcessAdditionalGid(uint32(group))
 	}
 	return nil
 }
 
+// generateUserString generates valid user string based on OCI Image Spec v1.0.0.
+func generateUserString(username, imageUser string, uid, gid *pb.Int64Value) (string, error) {
+	var userstr, groupstr string
+	if uid != nil {
+		userstr = strconv.FormatInt(uid.GetValue(), 10)
+	}
+	if username != "" {
+		userstr = username
+	}
+	if gid != nil {
+		groupstr = strconv.FormatInt(gid.GetValue(), 10)
+	}
+	// This may be not what kube expects and containerd/cri doesn't do it.
```

Comment:
We should retain falling back to imageUser.

Reply:
that's what I'm doing below, if no user/uid is present, we'll use imageUser which is still the image default.

Reply:
I'll remove this commnt also